Register array in RegisterBlock

kjetilkjeka opened the pull request:
I guess this pull request is dropping a bit out of the sky, but I realized I needed the feature sooner rather than later so decided to just go for it. When it's merged it should fix #114 


Some refactoring was done to allow for integration register arrays, the most significant are
- expand function only takes 1 register instead of slice. This means that the function has to be called for every register that should be expanded. This was done so a register can be conditionaly expanded depending on if it's possible to convert to a Rust array or not.
- register_to_rust method was factored out from register_block. Mainly so register block wouldn't grow very large with lots of nested control flow.
- ExpandedRegister was removed, the functions only use Register struct now. This was because if not it would require conversion functions when registers was not expanded. The struct would also need changing as it would only have to support arrays.
- Since the ExpandedRegisters was dropped the expand function now writes ```%s(i)``` (where i is the index) to the names of the expanded registers. This is used later on to resolve name and type.
- regex is a dependecy, this is used to replace ```%s(i)``` with the index. Adding this dependecy might not be desired? and if not I can reimplement it with some logic that iterates over the string.


I've tested with [S32K144.svd](https://github.com/kjetilkjeka/s32k144.rs/blob/master/src/S32K144.svd) and it seems to work perfectly with that file.

As mentioned in #114 i do hope that the method of using svd array information instead of unrolling arrays after is acceptable.

kjetilkjeka commented:
Also have a look at line 399 at generate.rs

I guess that is only valid as long as there are 4 byte (32 bit) registers. However I'm not sure how it can be generalized over register lengths?

japaric commented on src/generate.rs:
```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+                    let mut previous = 0;
+
+                    let get_number_option = |element_option: Option<&String>| -> Option<usize> {element_option
+                                                                        .and_then(|element| match element.parse::<usize>() {
```
this `match element.parse()` is equivalent to `element.parse().ok()`

japaric commented on src/generate.rs:
```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+                                                                        })
+                    };
+
+                    let mut index_convertible_temp = match get_number_option(index_iter.next()) {
```
this match is equivalent to `LHS == Some(0)`

japaric commented on src/generate.rs:
```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+                    for element in index_iter {
+                        if !index_convertible_temp { break; }
+
+                        index_convertible_temp = match get_number_option(Some(element)) {
```
this `get_number_option(Some(element))` would be simpler as `element.parse().ok()`

japaric replied:
the `match` can probably be replaced by `LHS == Some(previous + 1)`

japaric commented on src/generate.rs:
```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+        match *register {
+            Register::Single(ref _into) => registers_expanded.push(register.clone()),
+            Register::Array(ref _into, ref array_info) => {
+                let index_convertible = if let Some(ref indexes) = array_info.dim_index {
```
Could you add a comment about what this does? It wasn't entirely clear from reading the contents. Is it checking that `indexes == 0..info.dim`? In that case this may be simpler to write this as `indexes.iter().eq(0..info.dim)`.

japaric commented on src/generate.rs:
```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+                    false
+                };
+
+                let packed_registers = array_info.dim_increment == 4;
```
nit: given that this is a simpler check short circuiting on this equality may save some work

japaric commented on src/generate.rs:
```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+    let mut registers_expanded = vec![];
+
+    for register in registers {
+        match *register {
```
could you add a note here indicating that we'll only map a register array to an array type if there are no holes and each register in the array is 32 bits wide? Why is the 32 bits check required, by the way? Other registers arrays will be simply unrolled / expanded.

japaric commented on src/util.rs:
```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
+            let array_index_match = Regex::new(r"%s\((?P<i>[ a-zA-Z0-9]*)\)").unwrap();
+            //println!("{}", register.name.as_str());
+            let rty = Ident::new(array_index_match.replace_all(register.name.as_str(), "").to_sanitized_upper_case());
+            let reg_name = Ident::new(array_index_match.replace_all(register.name.as_str(), "$i").to_sanitized_snake_case());
```
As I asked elsewhere wouldn't it be better to replace the "%s" during `util::expand`?

japaric replied:
OK. I see why "%s(N)" is being used: to have expanded registers use the same type. This ... feels hackish; I would prefer if this was encoded in the type system rather than using a regex replace. ExpandedRegisters was used for this reason (to make all the expanded registers have the same type).

japaric commented on src/util.rs:
```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
 }
 
+
+pub fn sort_by_offset(mut registers: Vec<Register>) -> Vec<Register> {
```
I don't think this routine is complex enough that should be a function.

japaric commented on src/util.rs:
```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
-/// the register arrays have been expanded.
-pub fn expand(registers: &[Register]) -> Vec<ExpandedRegister> {
+/// and turn in into alist of registers where the register arrays have been expanded.
+pub fn expand(register: &Register) -> Vec<Register> {
```
Given that this will now only be used with `Register::Array` it may be better not to pass `&Register` but rather the contens of the `Register::Array` variant

japaric commented on src/util.rs:
```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
-                    info.name.replace("[%s]", "")
+            for (idx, i) in indices.iter().zip(0..) {
+                let name = if has_brackets {
+                    info.name.replace("[%s]", format!("%s({})", idx).as_str())
```
why do we want to keep the "%s" in this stage? Is it going to be used afterwards? Why not just directly replace it with the formatted `idx` at this point?

japaric commented on src/util.rs:
```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
+                    info.name.replace("[%s]", format!("%s({})", idx).as_str())
                 } else {
-                    info.name.replace("%s", "")
+                    info.name.replace("%s", format!("%s({})", idx).as_str())
```
same question as above

japaric commented on src/generate.rs:
```diff
@@ -381,36 +435,31 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+                    || {
+                        format!("Register {} has no `size` field", register.name)
+                    },
+                )? * array_info.dim / 8,
```
nit: I know that it's unrelated to this PR but could you replace the `8` by a `BITS_PER_BYTE` constant?

kjetilkjeka commented:
I agree that %s(N) is hacky. But the ExpandedRegister didn't work either as there was no support for array. I made a new struct extending syn::Field (by composition) adding size, offset and description. This makes perfect sense in my head as we're converting from svd::Register to rust fields (which syn::Field represents). Another bonus is the ability to use to_tokens instead of formating manually. Hope you agree with this reasoning? This made some refactoring necessary inside util but I think it makes more sense as it is now.

>>I guess that is only valid as long as there are 4 byte (32 bit) registers. However I'm not sure how it can be generalized over register lengths?

>I did not understand this constraint. Why are 4 byte registers special here? Shouldn't the check be register.size == dim_increment? That would check if each array element has reserved space apart from the rergiste

You're not suppose to understand that, was just me beeing stupid. Should be fixed now.

Most of your comments were considering the conditions for expanding. I've cleaned them up a whole lot. I think it's readable to the point making more comment than the one outside the function unnecessary. But since I wrote the code 2 hours ago I'm quite biased, say if you still want me to add a comment there.

I think I've addressed all the issues you mentioned except "Given that this will now only be used with Register::Array it may be better not to pass &Register but rather the contens of the Register::Array variant". I was a bit reluctant to change this out because:
-The similar function ```pub fn convert_svd_register(register: &svd::Register) -> syn::Field``` is used with svd::Register making my mental model of how to use these functions more consistent
- That expanding a Single register is allowed and gives exactly one registers makes sense (atleast in my head)
- One more argument gives the ability to make logic errors by passing non corresponding info and array_info
Anyway It's not a big deal but just think about it, if you still want it changed after, no problem.

homunkulus commented:
:umbrella: The latest upstream changes (presumably c7aed0a) made this pull request unmergeable. Please resolve the merge conflicts.

kjetilkjeka commented:
Rebase onto japaric/master to resolve merge conflicts and include bare_metal features

homunkulus commented:
:umbrella: The latest upstream changes (presumably 82ebef2) made this pull request unmergeable. Please resolve the merge conflicts.

kjetilkjeka commented:
Rebased onto 0.11.1 to make mergeable

kjetilkjeka commented:
@japaric What's the status on svd2rust at them moment? Now that the bare metal changes are adopted will this PR be next in line or should I expect others merged before it?

kjetilkjeka commented:
@japaric I guess you didn't have the time to look at this PR last time we talked about it.

Do you have an updated estimate of when you will be able to review this PR?

japaric commented on src/generate.rs:
```diff
@@ -461,69 +460,135 @@ pub fn peripheral(
 
+struct RegisterBlockField {
+    field: syn::Field,
+    description: Ident,
```
This is not a identifier; it should be a `String`.

japaric commented on src/generate.rs:
```diff
@@ -461,69 +460,135 @@ pub fn peripheral(
             "0x{:02x} - {}",
             register.offset,
-            util::respace(&register.info.description)
+            register.description,
```
Please revert this change or the documentation will have random newlines in it. As in: keep the `util::respace`.

kjetilkjeka commented:
Thank you for the review @japaric

> I've requested two changes: one is required to have svd2rust not regress in terms of documentation rendering. Please fix them and I'll merge this.

This should be fixed now.

> Do you know if any of the devices in the test suite is exercising this change? If none is could you add your SVD file to the test suite?

MK02F12810.svd should exercise this change (amongst other) for this register.

        <register>
          <dim>8</dim>
          <dimIncrement>0x4</dimIncrement>
          <dimIndex>0,1,2,3,4,5,6,7</dimIndex>
          <name>TAGVDW0S%s</name>
          <description>Cache Tag Storage</description>
          <addressOffset>0x100</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>
          <fields>
            <field>
              <name>valid</name>
              <description>1-bit valid for cache entry</description>
              <bitOffset>0</bitOffset>
              <bitWidth>1</bitWidth>
              <access>read-write</access>
            </field>
            <field>
              <name>tag</name>
              <description>14-bit tag for cache entry</description>
              <bitOffset>5</bitOffset>
              <bitWidth>14</bitWidth>
              <access>read-write</access>
            </field>
          </fields>
        </register>
        <register>

The diff for this register (between kjetilkjeka/svd2rust/master and kjetilkjeka/svd2rust/field_array) looks like the following:

    12303c12303,12317
    <         pub tagvdw0s: [TAGVDW0S; 8],
    ---
    >         pub tagvdw0s0: TAGVDW0S,
    >         #[doc = "0x104 - Cache Tag Storage"]
    >         pub tagvdw0s1: TAGVDW0S,
    >         #[doc = "0x108 - Cache Tag Storage"]
    >         pub tagvdw0s2: TAGVDW0S,
    >         #[doc = "0x10c - Cache Tag Storage"]
    >         pub tagvdw0s3: TAGVDW0S,
    >         #[doc = "0x110 - Cache Tag Storage"]
    >         pub tagvdw0s4: TAGVDW0S,
    >         #[doc = "0x114 - Cache Tag Storage"]
    >         pub tagvdw0s5: TAGVDW0S,
    >         #[doc = "0x118 - Cache Tag Storage"]
    >         pub tagvdw0s6: TAGVDW0S,
    >         #[doc = "0x11c - Cache Tag Storage"]
    >         pub tagvdw0s7: TAGVDW0S,

This means that the doc string is the same as it used to be for tagvdw0s0. "#[doc = 0x100 - Cache Tag Storage]"

I've also manually inspected the other registers in this file where the new code path is exercised and they are equivalent to the previous example.

The previous example is also equivalent to the .svd file I tested during devlopment [S32K144.svd](https://github.com/kjetilkjeka/s32k144.rs/blob/master/src/S32K144.svd)

I can, of course, manually inspect more files? But I've also tested it with the different (allowed) uses of dimIncrement and dimIndex manually. So I believe the next step should be to confirm that all the .svd files in the test suite will be compiled without error.

japaric commented:
Thanks, @kjetilkjeka.

@homunkulus r+

homunkulus commented:
:pushpin: Commit 52cf267 has been approved by `japaric`

<!-- @homunkulus r=japaric 52cf26750f435ddfb18719f70e711d9d903d570b -->

homunkulus commented:
:hourglass: Testing commit 52cf267 with merge 1f3542f...

homunkulus commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/japaric/svd2rust/build/1.0.122), [status-travis](https://travis-ci.org/japaric/svd2rust/builds/277468628?utm_source=github_status&utm_medium=notification)
Approved by: japaric
Pushing 1f3542f635c1512ca3155b1e0312021be2c08ee7 to master...
